java.lang.ClassNotFoundException: org.apache.logging.log4j.message.MessageFactory

**Precondition**
- [X] I checked the issues list for existing open or closed reports of the same problem.

**Describe the bug**
When running the latest Docker image (v12.1.8) in a GitLab CI pipeline, the application throws a java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/MessageFactory.

This exception appears to be triggered when the DataCacheFactory attempts to initialize the POM cache, which uses org.apache.commons.jcs3. This causes a secondary CacheException and results in the CentralAnalyzer's cache being disabled, as seen in the debug logs:

    DEBUG - Error constructing cache for POM files (followed by the NoClassDefFoundError)

    DEBUG - Setting: analyzer.central.use.cache='false'

    DEBUG - Error creating cache, disabling caching

This suggests a required log4j dependency is missing from the classpath in the latest Docker image.

**Version of dependency-check used**
The problem occurs using version 12.1.8 of the Docker image (CLI).

**Log file**
https://gist.github.com/bruville/8cabc5071e2f8070c8fe82743f2319ed

Key Exception Snippets:

```
2025-11-09 19:29:16,437 org.owasp.dependencycheck.data.cache.DataCacheFactory:164
DEBUG - Error constructing cache for POM files
java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/MessageFactory
    at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
    ...
    at org.apache.commons.jcs3.log.LogManager.getLog(LogManager.java:151)
    at org.apache.commons.jcs3.engine.control.CompositeCacheManager.<clinit>(CompositeCacheManager.java:80)
    at org.apache.commons.jcs3.JCS.getCacheManager(JCS.java:116)
    at org.apache.commons.jcs3.JCS.getInstance(JCS.java:159)
    at org.owasp.dependencycheck.data.cache.DataCacheFactory.getPomCache(DataCacheFactory.java:155)
    at org.owasp.dependencycheck.analyzer.CentralAnalyzer.initialize(CentralAnalyzer.java:124)
    ...
Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.message.MessageFactory
    ... 28 common frames omitted
```
  
Resulting in:

```
2025-11-09 19:29:16,438 org.owasp.dependencycheck.utils.Settings:1075
DEBUG - Setting: analyzer.central.use.cache='false'
2025-11-09 19:29:16,438 org.owasp.dependencycheck.analyzer.CentralAnalyzer:127
DEBUG - Error creating cache, disabling caching
org.apache.commons.jcs3.access.exception.CacheException: java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/MessageFactory
    at org.owasp.dependencycheck.data.cache.DataCacheFactory.getPomCache(DataCacheFactory.java:165)
    at org.owasp.dependencycheck.analyzer.CentralAnalyzer.initialize(CentralAnalyzer.java:124)
    ...
Caused by: java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/MessageFactory
    ...
Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.message.MessageFactory
    ... 28 common frames omitted

```

**To Reproduce**
Steps to reproduce the behavior:
1.     Run dependency-check using the owasp/dependency-check:latest image within a GitLab CI pipeline.
3.     Use the following .gitlab-ci.yml job definition, which includes entrypoint: [""] and the --log parameter (to capture debug output).
5.     Observe the scan.log file for the NoClassDefFoundError during initialization.

.gitlab-ci.yml Definition:
```
dependency_check:
  stage: dependency_scan
  image:
    name: owasp/dependency-check:latest
    entrypoint: [""] 

  allow_failure: true 
  
  variables:
    DATA_DIRECTORY: ".dependency-check-data"
    
  cache:
    key: dependency-check-db
    paths:
      - $DATA_DIRECTORY/

  before_script:
    - mkdir -p dependency-check-reports

  script:
    - |
      /usr/share/dependency-check/bin/dependency-check.sh \
        --project "MyProject" \
        --scan "." \
        --format "ALL" \
        --out "dependency-check-reports" \
        --nvdApiKey "$NVD_API_KEY" \
        --data "$DATA_DIRECTORY" \
        --enableExperimental \
        --log "dependency-check-reports/scan.log" \
        --ossIndexUsername "$OSSINDEX_USERNAME" \
        --ossIndexPassword "$OSSINDEX_PASSWORD"
```


**Expected behavior**
The scan should initialize all analyzers, including the CentralAnalyzer's POM cache, without throwing a NoClassDefFoundError. The cache should be enabled (analyzer.central.use.cache='true').

This is probably due to https://github.com/dependency-check/DependencyCheck/pull/7928 (JDK 21 -> 25) with root cause being a JDK regression in JDK 24/25 at https://bugs.openjdk.org/browse/JDK-8350481 - it has broken/regressed the JCS `ServiceLoader` mechanism used to configure loggers. Same regression has broken other libraries relying on `ServiceLoader` in the ecosystem (e.g [liquibase](https://github.com/liquibase/liquibase/issues/7318)).

Previously the `NoClassDefFoundError` would have been handled by `ServiceLoader`, rethrown as a `ServiceConfigurationError` which JCS handles, by rejecting Log4J as a valid `LogManager`.

https://github.com/apache/commons-jcs/blob/50773b22326312133d6bab63fc36a951bb795f54/commons-jcs-core/src/main/java/org/apache/commons/jcs3/log/LogManager.java#L56-L84

Can be replicated with  `org.apache.commons.jcs3.log.LogManager.getLog("test");` on JDK 25.

```
java.lang.NoClassDefFoundError: org/apache/logging/log4j/message/MessageFactory

	at java.base/java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredConstructors(Class.java:2985)
	at java.base/java.lang.Class.getConstructor0(Class.java:3180)
	at java.base/java.lang.Class.getConstructor(Class.java:2199)
	at java.base/java.util.ServiceLoader.getConstructor(ServiceLoader.java:623)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1111)
	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1142)
	at java.base/java.util.ServiceLoader$1.hasNext(ServiceLoader.java:1164)
	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1246)
	at org.apache.commons.jcs3.log.LogManager$LogFactoryHolder.createLogFactory(LogManager.java:69)
	at org.apache.commons.jcs3.log.LogManager$LogFactoryHolder.<clinit>(LogManager.java:48)
	at org.apache.commons.jcs3.log.LogManager.getLogFactory(LogManager.java:121)
	at org.apache.commons.jcs3.log.LogManager.getLog(LogManager.java:165)
	at org.owasp.dependencycheck.data.cache.DataCacheFactoryTest.testGetLogger(DataCacheFactoryTest.java:57)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1604)
Caused by: java.lang.ClassNotFoundException: org.apache.logging.log4j.message.MessageFactory
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	... 17 more
```

I think we can workaround this (and clean things up) by declaring our custom Slf4J provider directly.

--- a/ant/pom.xml
+++ b/ant/pom.xml
@@ -223,10 +223,6 @@ Copyright (c) 2013 - Jeremy Long. All Rights Reserved.
             <artifactId>slf4j-api</artifactId>
         </dependency>
         <dependency>
-            <groupId>io.github.jeremylong</groupId>
-            <artifactId>jcs3-slf4j</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.owasp</groupId>
             <artifactId>dependency-check-core</artifactId>
             <version>${project.parent.version}</version>

--- a/ant/src/main/java/org/owasp/dependencycheck/taskdefs/Purge.java
+++ b/ant/src/main/java/org/owasp/dependencycheck/taskdefs/Purge.java
@@ -17,7 +17,6 @@
  */
 package org.owasp.dependencycheck.taskdefs;
 
-import io.github.jeremylong.jcs3.slf4j.Slf4jAdapter;
 import java.io.File;
 import java.io.IOException;
 import java.io.InputStream;
@@ -128,9 +127,6 @@ public class Purge extends Task {
      * Hacky method of muting the noisy logging from JCS.
      */
     private void muteNoisyLoggers() {
-        System.setProperty("jcs.logSystem", "slf4j");
-        Slf4jAdapter.muteLogging(true);
-
         final String[] noisyLoggers = {
             "org.apache.hc"
         };

--- a/cli/pom.xml
+++ b/cli/pom.xml
@@ -67,7 +67,11 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
                     <archive>
                         <manifest>
                             <mainClass>org.owasp.dependencycheck.App</mainClass>
+                            <addClasspath>true</addClasspath>
                         </manifest>
+                        <manifestEntries>
+                            <Premain-Class>org.owasp.dependencycheck.PluginLoader</Premain-Class>
+                        </manifestEntries>
                     </archive>
                 </configuration>
             </plugin>
@@ -79,6 +83,11 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
                         <program>
                             <mainClass>org.owasp.dependencycheck.App</mainClass>
                             <id>dependency-check</id>
+                            <commandLineArguments>
+                                <commandLineArgument>-javaagent:@REPO@/${project.artifactId}-${project.version}.jar=@BASEDIR@/plugins</commandLineArgument>
+                                <commandLineArgument>-jar</commandLineArgument>
+                                <commandLineArgument>@REPO@/${project.artifactId}-${project.version}.jar</commandLineArgument>
+                            </commandLineArguments>
                         </program>
                     </programs>
                     <assembleDirectory>${project.build.directory}/release</assembleDirectory>
@@ -88,10 +97,8 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
                     </binFileExtensions>
                     <repositoryLayout>flat</repositoryLayout>
                     <repositoryName>lib</repositoryName>
-                    <useWildcardClassPath>true</useWildcardClassPath>
-                    <configurationDirectory>plugins/*</configurationDirectory>
-                    <includeConfigurationDirectoryInClasspath>true</includeConfigurationDirectoryInClasspath>
-                    <unixScriptTemplate>${project.basedir}/src/main/conf/unixBinTemplate</unixScriptTemplate>
+                    <unixScriptTemplate>${project.basedir}/src/main/conf/unixBinTemplate.sh</unixScriptTemplate>
+                    <windowsScriptTemplate>${project.basedir}/src/main/conf/windowsBinTemplate.bat</windowsScriptTemplate>
                     <!--
                        enable-native-access=ALL-UNNAMED
                             Java 21+: Needed by Lucene indexes unless we do -Dorg.apache.lucene.store.MMapDirectory.enableMemorySegments=false
@@ -107,25 +114,6 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
                         <goals>
                             <goal>assemble</goal>
                         </goals>
-                    </execution>
-                </executions>
-            </plugin>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-antrun-plugin</artifactId>
-                <executions>
-                    <execution>
-                        <id>fix-windows-shell-script</id>
-                        <phase>package</phase>
-                        <goals>
-                            <goal>run</goal>
-                        </goals>
-                        <configuration>
-                            <!-- Hack/workaround for https://github.com/mojohaus/appassembler/issues/114 -->
-                            <target>
-                                <replace file="${project.build.directory}/release/bin/dependency-check.bat" token="%JAVACMD% %JAVA_OPTS%" value="&quot;%JAVACMD%&quot; %JAVA_OPTS%" failOnNoReplacements="true" />
-                            </target>
-                        </configuration>
                     </execution>
                 </executions>
             </plugin>
@@ -197,6 +185,11 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
                     <artifactId>tools</artifactId>
                 </exclusion>
             </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
         </dependency>
     </dependencies>
 </project>

--- a/cli/pom.xml
+++ b/cli/pom.xml
@@ -180,10 +180,6 @@ Copyright (c) 2012 - Jeremy Long. All Rights Reserved.
             <artifactId>slf4j-api</artifactId>
         </dependency>
         <dependency>
-            <groupId>io.github.jeremylong</groupId>
-            <artifactId>jcs3-slf4j</artifactId>
-        </dependency>
-        <dependency>
             <!-- not visible in imports due to method chaining, but App code uses classes from this library -->
             <groupId>io.github.jeremylong</groupId>
             <artifactId>open-vulnerability-clients</artifactId>

--- a/cli/src/main/conf/unixBinTemplate.sh
+++ b/cli/src/main/conf/unixBinTemplate.sh
@@ -53,7 +53,6 @@ fi
 # For Cygwin and MINGW, ensure paths are in UNIX format before anything is touched
 if $cygwin || $mingw; then
   [ -n "$JAVA_HOME" ] && JAVA_HOME=`cygpath --unix "$JAVA_HOME"`
-  [ -n "$CLASSPATH" ] && CLASSPATH=`cygpath --path --unix "$CLASSPATH"`
 fi
 
 # If a specific java binary isn't specified search for the standard 'java' binary
@@ -81,20 +80,8 @@ then
   REPO="$BASEDIR"/@REPO@
 fi
 
-CLASSPATH=@CLASSPATH@
-
-ENDORSED_DIR=@ENDORSED_DIR@
-if [ -n "$ENDORSED_DIR" ] ; then
-  CLASSPATH=$BASEDIR/$ENDORSED_DIR/*:$CLASSPATH
-fi
-
-if [ -n "$CLASSPATH_PREFIX" ] ; then
-  CLASSPATH=$CLASSPATH_PREFIX:$CLASSPATH
-fi
-
 # For Cygwin and Mingw, switch paths to Windows format before running java
 if $cygwin || $mingw; then
-  [ -n "$CLASSPATH" ] && CLASSPATH=`cygpath --path --windows "$CLASSPATH"`
   [ -n "$JAVA_HOME" ] && JAVA_HOME=`cygpath --path --windows "$JAVA_HOME"`
   [ -n "$HOME" ] && HOME=`cygpath --path --windows "$HOME"`
   [ -n "$BASEDIR" ] && BASEDIR=`cygpath --path --windows "$BASEDIR"`
@@ -110,11 +97,9 @@ fi
 done
 
 exec "$JAVACMD" $JAVA_OPTS $DEBUG @EXTRA_JVM_ARGUMENTS@ \
-  -classpath "$CLASSPATH" \
   -Dapp.name="@APP_NAME@" \
   -Dapp.pid="$$" \
   -Dapp.repo="$REPO" \
   -Dapp.home="$BASEDIR" \
   -Dbasedir="$BASEDIR" \
-  @MAINCLASS@ \
   @APP_ARGUMENTS@"$@"@UNIX_BACKGROUND@

--- a/cli/src/main/conf/windowsBinTemplate.bat
+++ b/cli/src/main/conf/windowsBinTemplate.bat
@@ -1,0 +1,88 @@
+#LICENSE_HEADER#
+@echo off
+
+set ERROR_CODE=0
+
+:init
+@REM Decide how to startup depending on the version of windows
+
+@REM -- Win98ME
+if NOT "%OS%"=="Windows_NT" goto Win9xArg
+
+@REM set local scope for the variables with windows NT shell
+if "%OS%"=="Windows_NT" @setlocal
+
+@REM -- 4NT shell
+if "%eval[2+2]" == "4" goto 4NTArgs
+
+@REM -- Regular WinNT shell
+set CMD_LINE_ARGS=%*
+goto WinNTGetScriptDir
+
+@REM The 4NT Shell from jp software
+:4NTArgs
+set CMD_LINE_ARGS=%$
+goto WinNTGetScriptDir
+
+:Win9xArg
+@REM Slurp the command line arguments.  This loop allows for an unlimited number
+@REM of arguments (up to the command line limit, anyway).
+set CMD_LINE_ARGS=
+:Win9xApp
+if %1a==a goto Win9xGetScriptDir
+set CMD_LINE_ARGS=%CMD_LINE_ARGS% %1
+shift
+goto Win9xApp
+
+:Win9xGetScriptDir
+set SAVEDIR=%CD%
+%0\
+cd %0\..\..
+set BASEDIR=%CD%
+cd %SAVEDIR%
+set SAVE_DIR=
+goto repoSetup
+
+:WinNTGetScriptDir
+for %%i in ("%~dp0..") do set "BASEDIR=%%~fi"
+
+:repoSetup
+set REPO=
+#ENV_SETUP#
+
+if "%JAVACMD%"=="" set JAVACMD=#JAVA_BINARY#
+
+if "%REPO%"=="" set REPO=%BASEDIR%\#REPO#
+
+@REM Reaching here means variables are defined and arguments have been captured
+:endInit
+
+"%JAVACMD%" %JAVA_OPTS% #EXTRA_JVM_ARGUMENTS# -Dapp.name="#APP_NAME#" -Dapp.repo="%REPO%" -Dapp.home="%BASEDIR%" -Dbasedir="%BASEDIR%" #APP_ARGUMENTS#%CMD_LINE_ARGS%
+if %ERRORLEVEL% NEQ 0 goto error
+goto end
+
+:error
+if "%OS%"=="Windows_NT" @endlocal
+set ERROR_CODE=%ERRORLEVEL%
+
+:end
+@REM set local scope for the variables with windows NT shell
+if "%OS%"=="Windows_NT" goto endNT
+
+@REM For old DOS remove the set variables from ENV - we assume they were not set
+@REM before we started - at least we don't leave any baggage around
+set CMD_LINE_ARGS=
+goto postExec
+
+:endNT
+@REM If error code is set to 1 then the endlocal was done already in :error.
+if %ERROR_CODE% EQU 0 @endlocal
+
+
+:postExec
+
+if "%FORCE_EXIT_ON_ERROR%" == "on" (
+  if %ERROR_CODE% NEQ 0 exit %ERROR_CODE%
+)
+
+exit /B %ERROR_CODE%

--- a/cli/src/main/java/org/owasp/dependencycheck/App.java
+++ b/cli/src/main/java/org/owasp/dependencycheck/App.java
@@ -48,7 +48,6 @@ import ch.qos.logback.classic.filter.ThresholdFilter;
 import ch.qos.logback.classic.spi.ILoggingEvent;
 import ch.qos.logback.classic.Level;
 import ch.qos.logback.classic.LoggerContext;
-import io.github.jeremylong.jcs3.slf4j.Slf4jAdapter;
 import java.util.TreeSet;
 import org.owasp.dependencycheck.utils.SeverityUtil;
 
@@ -84,10 +83,6 @@ public class App {
      */
     @SuppressWarnings("squid:S4823")
     public static void main(String[] args) {
-        System.setProperty("jcs.logSystem", "slf4j");
-        if (!LOGGER.isDebugEnabled()) {
-            Slf4jAdapter.muteLogging(true);
-        }
         final int exitCode;
         final App app = new App();
         exitCode = app.run(args);

--- a/cli/src/main/java/org/owasp/dependencycheck/PluginLoader.java
+++ b/cli/src/main/java/org/owasp/dependencycheck/PluginLoader.java
@@ -1,0 +1,38 @@
+package org.owasp.dependencycheck;
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.instrument.Instrumentation;
+import java.util.jar.JarFile;
+
+/**
+ * Java agent for loading plugin JARs from a specified directory into the system classpath
+ * before the main application starts. This allows additional plugins to be available at runtime
+ * by appending their JAR files to the system class loader search path; while allowing use of an
+ * executable jar with deterministic classpath ordering.
+ * <p/>
+ * To use, specify this class as a Java agent and provide the plugins directory as the -javaagent argument
+ */
+public class PluginLoader {
+    /**
+     * Java agent entry point. Loads all JAR files from the specified plugins directory
+     * and appends them to the system class loader search path.
+     *
+     * @param agentArg the path to the plugins directory containing JAR files to load, e.g `-javaagent:cli.jar=/usr/share/dependency-check/plugins`
+     * @param inst     the instrumentation instance provided by the JVM
+     */
+    public static void premain(String agentArg, Instrumentation inst) {
+        File pluginsDir = new File(agentArg);
+        if (pluginsDir.isDirectory()) {
+            File[] files = pluginsDir.listFiles((dir, name) -> name.endsWith(".jar"));
+            for (File file : files == null ? new File[0] : files) {
+                try (JarFile jar = new JarFile(file)) {
+                    inst.appendToSystemClassLoaderSearch(jar);
+                } catch (IOException e) {
+                    System.err.printf("[WARN] Failed to read plugin jar file at %s. Jar will not be available on classpath: %s%n", file, e);
+                    e.printStackTrace(System.err);
+                }
+            }
+        }
+    }
+}

--- a/cli/src/main/resources/logback.xml
+++ b/cli/src/main/resources/logback.xml
@@ -11,7 +11,7 @@
     </appender>
 
     <logger name="org.apache.lucene" level="ERROR" />
-    <logger name="org.apache.commons.jcs" level="ERROR" />
+    <logger name="org.apache.commons.jcs3" level="FATAL" />
     <logger name="org.apache.hc" level="ERROR" />
 
     <root level="INFO">

--- a/cli/src/test/java/org/owasp/dependencycheck/PluginLoaderTest.java
+++ b/cli/src/test/java/org/owasp/dependencycheck/PluginLoaderTest.java
@@ -1,0 +1,72 @@
+package org.owasp.dependencycheck;
+
+import org.apache.commons.compress.archivers.zip.ZipArchiveOutputStream;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.mockito.Mockito;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.lang.instrument.Instrumentation;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.regex.Pattern;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.matchesPattern;
+import static org.mockito.ArgumentMatchers.assertArg;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+
+class PluginLoaderTest {
+
+    private final Instrumentation instrumentation = Mockito.mock(Instrumentation.class);
+
+    @TempDir
+    Path tempDir;
+
+    @Test
+    void shouldDoNothingIfDirectoryDoesntExist() {
+        PluginLoader.premain("blah", instrumentation);
+        verifyNoInteractions(instrumentation);
+    }
+
+    @Test
+    void shouldDoNothingIfDirectoryIsEmpty() {
+        PluginLoader.premain(tempDir.toString(), instrumentation);
+        verifyNoInteractions(instrumentation);
+    }
+
+    @Test
+    void shouldAddJarToClassPath() throws Exception {
+        createEmptyValidJar();
+        createEmptyValidJar();
+        PluginLoader.premain(tempDir.toString(), instrumentation);
+        verify(instrumentation, times(2))
+                .appendToSystemClassLoaderSearch(assertArg(jar -> assertThat(jar.getName(), matchesPattern(".*/dummy.*\\.jar"))));
+    }
+
+    @Test
+    void shouldStopLoadingPluginsOnBadJarButSucceed() throws Exception {
+        PrintStream originalErr = System.err;
+        ByteArrayOutputStream errContent = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(errContent));
+        try {
+            createEmptyBadJar();
+            PluginLoader.premain(tempDir.toString(), instrumentation);
+            assertThat(errContent.toString(), matchesPattern(Pattern.compile("\\[WARN\\] Failed to read plugin jar file at .*/dummy.*\\.jar\\. Jar will not be available on classpath.*zip file is empty.*", Pattern.DOTALL)));
+        } finally {
+            System.setErr(originalErr);
+        }
+    }
+
+    private Path createEmptyBadJar() throws IOException {
+        return Files.createTempFile(tempDir, "dummy", ".jar");
+    }
+
+    private void createEmptyValidJar() throws IOException {
+        new ZipArchiveOutputStream(createEmptyBadJar().toFile()).close();
+    }
+}

--- a/core/pom.xml
+++ b/core/pom.xml
@@ -197,14 +197,14 @@ Copyright (c) 2012 Jeremy Long. All Rights Reserved.
             <groupId>org.whitesource</groupId>
             <artifactId>pecoff4j</artifactId>
         </dependency>
-        <dependency>
-            <groupId>org.apache.commons</groupId>
-            <artifactId>commons-jcs3-core</artifactId>
-        </dependency>
+        <!-- JCS 3 logging adapter; intentionally higher on classpath than JCS itself -->
         <dependency>
             <groupId>io.github.jeremylong</groupId>
             <artifactId>jcs3-slf4j</artifactId>
-            <scope>runtime</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-jcs3-core</artifactId>
         </dependency>
         <dependency>
             <groupId>com.github.package-url</groupId>

--- a/core/src/main/java/org/owasp/dependencycheck/data/cache/DataCache.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/cache/DataCache.java
@@ -38,7 +38,7 @@ public class DataCache<T> {
      *
      * @param cache a reference to the underlying cache implementation.
      */
-    public DataCache(CacheAccess<String, T> cache) {
+    DataCache(CacheAccess<String, T> cache) {
         this.cache = cache;
     }
 

--- a/core/src/main/java/org/owasp/dependencycheck/data/cache/DataCacheFactory.java
+++ b/core/src/main/java/org/owasp/dependencycheck/data/cache/DataCacheFactory.java
@@ -22,6 +22,8 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.util.List;
 import java.util.Properties;
+
+import io.github.jeremylong.jcs3.slf4j.Slf4jAdapter;
 import org.apache.commons.jcs3.JCS;
 import org.apache.commons.jcs3.access.CacheAccess;
 import org.apache.commons.jcs3.access.exception.CacheException;
@@ -75,6 +77,13 @@ public class DataCacheFactory {
          * Used to store POM files retrieved from central.
          */
         POM
+    }
+
+    static {
+        System.setProperty("jcs.logSystem", "slf4j");
+        if (!LOGGER.isTraceEnabled()) {
+            Slf4jAdapter.muteLogging(true);
+        }
     }
 
     /**

--- a/maven/pom.xml
+++ b/maven/pom.xml
@@ -125,10 +125,6 @@ Copyright (c) 2013 Jeremy Long. All Rights Reserved.
             <scope>test</scope>
         </dependency>
         <dependency>
-            <groupId>io.github.jeremylong</groupId>
-            <artifactId>jcs3-slf4j</artifactId>
-        </dependency>
-        <dependency>
             <groupId>org.apache.maven</groupId>
             <artifactId>maven-plugin-api</artifactId>
             <scope>provided</scope>

--- a/maven/src/main/java/org/owasp/dependencycheck/maven/BaseDependencyCheckMojo.java
+++ b/maven/src/main/java/org/owasp/dependencycheck/maven/BaseDependencyCheckMojo.java
@@ -20,7 +20,6 @@ package org.owasp.dependencycheck.maven;
 import com.github.packageurl.MalformedPackageURLException;
 import com.github.packageurl.PackageURL.StandardTypes;
 import com.github.packageurl.PackageURL;
-import io.github.jeremylong.jcs3.slf4j.Slf4jAdapter;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -2731,11 +2730,6 @@ public abstract class BaseDependencyCheckMojo extends AbstractMojo implements Ma
      * Hacky method of muting the noisy logging from JCS
      */
     protected void muteNoisyLoggers() {
-        System.setProperty("jcs.logSystem", "slf4j");
-        if (!getLog().isDebugEnabled()) {
-            Slf4jAdapter.muteLogging(true);
-        }
-
         final String[] noisyLoggers = {
             "org.apache.hc"
         };

